nodeJS/express/using_postgresql.md

```diff
@@ -0,0 +1,309 @@
+  port: 5432 // The default port
+});
+```
+
+An alternative to defining the connection information is through a [Connection URI](https://node-postgres.com/features/connecting#connection-uri). You'll likely be using connection URIs when connecting with a hosted database service. Here's what it would look like based on the above properties:
+
+```javascript
+const { Pool } = require("pg");
+
+// Again, this should be read from an environment variable
+module.exports = new Pool({
+  connectionString: "postgresql://<role_name>:<role_password>@localhost:5432/top_users"
+});
+```
+
+Pick whichever method you want to use and let's continue.
+
+<div class="lesson-note" markdown="1">
+
+#### Two ways of connecting with pg
+
+`pg` has two ways to connect to a db: a client and a pool.
+
+Client is an individual connection to the DB, which you manually manage. You open a connection, do your query, then close it. This is fine for one-off queries, but can become expensive if you're dealing with a lot of queries. Wouldn't this problem be alleviated if we could somehow hold onto a client? Yes!
+
+Enter pool. As the name suggests, it's a pool of clients. A pool holds onto connections. And when you query, it'll programmatically open a new connection unless there's an existing spare one. Perfect for web servers.
+
+</div>
+
+### Querying with pg
+
+With our initialized `Pool`, we can use the `query` method. Create a new `db/queries.js` file. Upon revising our project requirements, we understand we need two db interactions: getting all usernames and inserting a new username. Let's define these functions:
+
+```javascript
+const pool = require("./pool");
+
+async function getAllUsernames() {
+  const { rows } = await db.query("SELECT * FROM usernames");
+  return rows;
+}
+
+async function insertUsername(username) {
+  await db.query("INSERT INTO usernames (username) VALUES ($1)", [username]);
+}
+
+module.exports = {
+  getAllUsernames,
+  insertUsername
+};
+```
+
+<div class="lesson-note" markdown="1">
+
+#### Parameterization
+
+What's with the `$1` in the insert query?
+
+Alternatively, the query could look like:
+
+```javascript
+await db.query("INSERT INTO usernames (username) VALUES ('" + username + "')");
+```
+
+We're passing user entered value i.e. `username` directly into our query. A nefarious user could enter something like `sike'); DROP TABLE usernames; --` and wreak havoc. Scary stuff. This is called [SQL injection](https://en.wikipedia.org/wiki/SQL_injection).
+
+`pg` provides [query parameterization](https://node-postgres.com/features/queries#parameterized-query) to prevent this. Instead of passing user input directly, we pass it in an array as the second argument. `pg` handles the rest.
+
+</div>
+
+Invoke the above two functions in the specific controllers (you might have different function names etc. The important thing is to understand how the db functions are invoked):
+
+```javascript
+const db = require("../db/queries");
```

Comment:
I've commented above that named the pool module as `db` should we name this as `query` instead

```javascript
await query.getAllUsernames();

await query.insertUsername(username);
```
Or we can make it more modular by creating `db/queries/user.js` and have generic names like `Username.getAll()` `Username.insert()` but it's up to you

What do you think?

Reply:
I'd rather keep it as `db`, though at the end of the day I really don't feel strongly towards it. I've updated the other code example., doo let me know what you prefer though after you're read that one; I'll update the lesson accordingly.

Regarding your suggestion on making it modular: I'll trust the learners to refactor as they see fit when they work on bigger projects. They already have all the knowledge they need to do `Username.getAll()` or however else they want to refactor.

Reply:
^ adding to the above, you might've also noticed how barebones the script was compared to the version you wrote (which included utility functions etc). I'd rather have learners come up with this stuff